lib/pbench/agent/meister/__init__.py

```diff
@@ -0,0 +1,800 @@
+        independent of a ToolMeister implementation, but we keep this method
+        in the ToolMeister class since it is closely related to the
+        implementation.
+        """
+        try:
+            benchmark_run_dir = params["benchmark_run_dir"]
+            channel = params["channel"]
+            controller = params["controller"]
+            group = params["group"]
+            hostname = params["hostname"]
+            tools = params["tools"]
+        except KeyError as exc:
+            raise ToolMeisterError(f"Invalid parameter block, missing key {exc}")
+        else:
+            return benchmark_run_dir, channel, controller, group, hostname, tools
+
+    def __init__(self, pbench_bin, params, redis_server, logger):
+        self.logger = logger
+        self.pbench_bin = pbench_bin
+        ret_val = self.fetch_params(params)
+        (
+            self._benchmark_run_dir,
+            self._channel,
+            self._controller,
+            self._group,
+            self._hostname,
+            self._tools,
+        ) = ret_val
+        self._running_tools = dict()
+        self._rs = redis_server
+        logger.debug("pubsub")
+        self._pubsub = self._rs.pubsub()
+        logger.debug("subscribe %s", self._channel)
+        self._pubsub.subscribe(self._channel)
+        logger.debug("listen")
+        self._chan = self._pubsub.listen()
+        logger.debug("done listening")
+        # Now that we have subscribed to the channel as specified in the
+        # params object, we need to pull off the first message, which is an
+        # acknowledgement that we have properly subscribed.
+        logger.debug("next")
+        resp = next(self._chan)
+        assert resp["type"] == "subscribe", f"Unexpected 'type': {resp!r}"
+        assert resp["pattern"] is None, f"Unexpected 'pattern': {resp!r}"
+        assert (
+            resp["channel"].decode("utf-8") == self._channel
+        ), f"Unexpected 'channel': {resp!r}"
+        assert resp["data"] == 1, f"Unexpected 'data': {resp!r}"
+        logger.debug("next done")
+        # We start in the "idle" state.
+        self.state = "idle"
+        self._valid_states = frozenset(["idle", "running"])
+        self._state_trans = {
+            "start": {"curr": "idle", "next": "running", "action": self.start_tools},
+            "stop": {"curr": "running", "next": "idle", "action": self.stop_tools},
+        }
+        self._valid_actions = frozenset(["start", "stop", "send", "terminate"])
+        for key in self._state_trans.keys():
+            assert (
+                key in self._valid_actions
+            ), f"INTERNAL ERROR: invalid state transition entry, '{key}'"
+            assert self._state_trans[key]["next"] in self._valid_states, (
+                "INTERNAL ERROR: invalid state transition 'next' entry for"
+                f" '{key}', '{self._state_trans[key]['next']}'"
+            )
```

Comment:
I think since the values of `self._state_trans`, `self._valid_actions`, and `self._valid_states`  variables are not changing, this for loop is gonna be always true so why do we need to assert that?

Reply:
straight copy of toolmeister